<div style="direction:rtl;line-height:300%;text-align:justify;" align="justify"><font face="B nazanin" size=4>
    <b style="font-size: 30px">مقدمه</b><br>
پیش از تکمیل بخش‌های زیر توجه کنید که در هربخش, قسمت‌هایی تحت عنوان 
    <b>"ارزیابی دستیار آموزشی"</b> وجود دارد که بایستی پس از طی کردن فرآیند‌های آموزش و تست,در خروجی نتیجه‌ی ارزیابی متریک‌های خواسته شده در آن قسمت از سوال, فارغ از دانستن داده‌ی تست (که در اختیار شما قرار نگرفته است) را چاپ کنید.
</font></div>

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import copy #if needed!
from numpy import *
import math


%matplotlib inline

In [2]:
#for TAs

#data_path_TA = ""
#df_test_TA = pd.read_csv(data_path_TA)
#X_TA = df_test_ta.drop(colums='target')
#y_TA = df_test_ta['target']

<div style="direction:ltr;line-height:300%;text-align:justify;" align="justify"><font face="B nazanin" size=4>
<b style="font-size: 35px">Bayes, Naiive, Logistic</b>
<br>
</font></div>

<div style="direction:rtl;line-height:300%;text-align:justify;" align="justify"><font face="B nazanin" size=4>
در این قسمت توابعی را تکمیل خواهید کرد که در ادامه با استفاده از آن ها بخش های سوال اول را پاسخ خواهید داد.
<br>
    نکته: به صورت کلی توابع باید عملکرد های گفته شده را داشته باشند اما با توجه به نیاز می‌توانید ورودی ها را مطابق میل تغییر دهید.
</font></div>

### 1

In [3]:
data_base = pd.read_csv("heart_train.csv")
print(data_base) 

     Unnamed: 0  age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  \
0             0   57    1   2       150   168    0        1      174      0   
1             1   46    1   1       101   197    1        1      156      0   
2             2   54    0   2       135   304    1        1      170      0   
3             3   48    1   1       110   229    0        1      168      0   
4             4   58    0   0       130   197    0        1      131      0   
..          ...  ...  ...  ..       ...   ...  ...      ...      ...    ...   
696         696   38    1   2       138   175    0        1      173      0   
697         697   47    1   2       108   243    0        1      152      0   
698         698   61    0   0       145   307    0        0      146      1   
699         699   57    0   0       128   303    0        0      159      0   
700         700   60    1   0       130   253    0        1      144      1   

     oldpeak  slope  ca  thal  target  
0        1.

In [4]:
data_base.drop(["Unnamed: 0"], axis = 1, inplace = True)
data_base

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     57    1   2       150   168    0        1      174      0      1.6   
1     46    1   1       101   197    1        1      156      0      0.0   
2     54    0   2       135   304    1        1      170      0      0.0   
3     48    1   1       110   229    0        1      168      0      1.0   
4     58    0   0       130   197    0        1      131      0      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
696   38    1   2       138   175    0        1      173      0      0.0   
697   47    1   2       108   243    0        1      152      0      0.0   
698   61    0   0       145   307    0        0      146      1      1.0   
699   57    0   0       128   303    0        0      159      0      0.0   
700   60    1   0       130   253    0        1      144      1      1.4   

     slope  ca  thal  target  
0        2   0     2       1  
1        2   0     3       1  
2        2   0     2       1  
3        0   0     3       0  
4        1   0     2       1  
..     ...  ..   ...     ...  
696      2   4     2       1  
697      2   0     2       0  
698      1   0     3       0  
699      2   1     2       1  
700      2   1     3       0  

[701 rows x 14 columns]

In [5]:
header=data_base.columns
header

Index(['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach',
       'exang', 'oldpeak', 'slope', 'ca', 'thal', 'target'],
      dtype='object')

In [6]:
list_discrete_no=[]
list_discrete_tag=[]
for i in range(len(header)):
    if (len(np.unique(data_base[header[i]]))<=5):
        #print(header[i])
        list_discrete_tag.append(header[i])
        #print(np.unique(data_base[header[i]]))
        list_discrete_no.append(np.unique(data_base[header[i]]))

In [7]:
list_discrete_tag

['sex', 'cp', 'fbs', 'restecg', 'exang', 'slope', 'ca', 'thal', 'target']

In [8]:
data_base_disceret = pd.DataFrame(data_base,columns=list_discrete_tag)
data_base_disceret

sex  cp  fbs  restecg  exang  slope  ca  thal  target
0      1   2    0        1      0      2   0     2       1
1      1   1    1        1      0      2   0     3       1
2      0   2    1        1      0      2   0     2       1
3      1   1    0        1      0      0   0     3       0
4      0   0    0        1      0      1   0     2       1
..   ...  ..  ...      ...    ...    ...  ..   ...     ...
696    1   2    0        1      0      2   4     2       1
697    1   2    0        1      0      2   0     2       0
698    0   0    0        0      1      1   0     3       0
699    0   0    0        0      0      2   1     2       1
700    1   0    0        1      1      2   1     3       0

[701 rows x 9 columns]

In [9]:
list_discrete_no

[array([0, 1], dtype=int64),
 array([0, 1, 2, 3], dtype=int64),
 array([0, 1], dtype=int64),
 array([0, 1, 2], dtype=int64),
 array([0, 1], dtype=int64),
 array([0, 1, 2], dtype=int64),
 array([0, 1, 2, 3, 4], dtype=int64),
 array([0, 1, 2, 3], dtype=int64),
 array([0, 1], dtype=int64)]

In [10]:
gp=data_base.groupby(by='target').size()
gp

target
0    348
1    353
dtype: int64

In [11]:
data_base_0=data_base[data_base['target']==0]
data_base_0

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
3     48    1   1       110   229    0        1      168      0      1.0   
5     51    0   0       130   305    0        1      142      1      1.2   
6     61    1   0       140   207    0        0      138      1      1.9   
8     62    1   0       120   267    0        1       99      1      1.8   
10    62    0   0       160   164    0        0      145      0      6.2   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
689   59    1   0       170   326    0        0      140      1      3.4   
691   57    1   0       130   131    0        1      115      1      1.2   
697   47    1   2       108   243    0        1      152      0      0.0   
698   61    0   0       145   307    0        0      146      1      1.0   
700   60    1   0       130   253    0        1      144      1      1.4   

     slope  ca  thal  target  
3        0   0     3       0  
5        1   0     3       0  
6        2   1     3       0  
8        1   2     3       0  
10       0   3     3       0  
..     ...  ..   ...     ...  
689      0   0     3       0  
691      1   1     3       0  
697      2   0     2       0  
698      1   0     3       0  
700      2   1     3       0  

[348 rows x 14 columns]

In [12]:
data_base_1=data_base[data_base['target']==1]
data_base_1

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     57    1   2       150   168    0        1      174      0      1.6   
1     46    1   1       101   197    1        1      156      0      0.0   
2     54    0   2       135   304    1        1      170      0      0.0   
4     58    0   0       130   197    0        1      131      0      0.6   
7     34    1   3       118   182    0        0      174      0      0.0   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
693   38    1   2       138   175    0        1      173      0      0.0   
694   47    1   2       138   257    0        0      156      0      0.0   
695   67    0   2       115   564    0        0      160      0      1.6   
696   38    1   2       138   175    0        1      173      0      0.0   
699   57    0   0       128   303    0        0      159      0      0.0   

     slope  ca  thal  target  
0        2   0     2       1  
1        2   0     3       1  
2        2   0     2       1  
4        1   0     2       1  
7        2   0     2       1  
..     ...  ..   ...     ...  
693      2   4     2       1  
694      2   0     2       1  
695      1   0     3       1  
696      2   4     2       1  
699      2   1     2       1  

[353 rows x 14 columns]

In [13]:
for i in range(len(list_discrete_tag)-1):
    data_base_0.drop(list_discrete_tag[i],axis=1,inplace=True)
data_base_0

C:\Users\asus\anaconda3\lib\site-packages\pandas\core\frame.py:4163: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


age  trestbps  chol  thalach  oldpeak  target
3     48       110   229      168      1.0       0
5     51       130   305      142      1.2       0
6     61       140   207      138      1.9       0
8     62       120   267       99      1.8       0
10    62       160   164      145      6.2       0
..   ...       ...   ...      ...      ...     ...
689   59       170   326      140      3.4       0
691   57       130   131      115      1.2       0
697   47       108   243      152      0.0       0
698   61       145   307      146      1.0       0
700   60       130   253      144      1.4       0

[348 rows x 6 columns]

In [14]:
y0=data_base_0['target']
y0

3      0
5      0
6      0
8      0
10     0
      ..
689    0
691    0
697    0
698    0
700    0
Name: target, Length: 348, dtype: int64

In [15]:
data_base_0.drop('target',axis=1,inplace=True)
data_base_0

C:\Users\asus\anaconda3\lib\site-packages\pandas\core\frame.py:4163: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


age  trestbps  chol  thalach  oldpeak
3     48       110   229      168      1.0
5     51       130   305      142      1.2
6     61       140   207      138      1.9
8     62       120   267       99      1.8
10    62       160   164      145      6.2
..   ...       ...   ...      ...      ...
689   59       170   326      140      3.4
691   57       130   131      115      1.2
697   47       108   243      152      0.0
698   61       145   307      146      1.0
700   60       130   253      144      1.4

[348 rows x 5 columns]

In [16]:
for i in range(len(list_discrete_tag)-1):
    data_base_1.drop(list_discrete_tag[i],axis=1,inplace=True)

C:\Users\asus\anaconda3\lib\site-packages\pandas\core\frame.py:4163: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


In [17]:
data_base_1

age  trestbps  chol  thalach  oldpeak  target
0     57       150   168      174      1.6       1
1     46       101   197      156      0.0       1
2     54       135   304      170      0.0       1
4     58       130   197      131      0.6       1
7     34       118   182      174      0.0       1
..   ...       ...   ...      ...      ...     ...
693   38       138   175      173      0.0       1
694   47       138   257      156      0.0       1
695   67       115   564      160      1.6       1
696   38       138   175      173      0.0       1
699   57       128   303      159      0.0       1

[353 rows x 6 columns]

In [18]:
y1=data_base_1['target']
y1

0      1
1      1
2      1
4      1
7      1
      ..
693    1
694    1
695    1
696    1
699    1
Name: target, Length: 353, dtype: int64

In [19]:
data_base_1.drop('target',axis=1,inplace=True)
data_base_1

C:\Users\asus\anaconda3\lib\site-packages\pandas\core\frame.py:4163: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


age  trestbps  chol  thalach  oldpeak
0     57       150   168      174      1.6
1     46       101   197      156      0.0
2     54       135   304      170      0.0
4     58       130   197      131      0.6
7     34       118   182      174      0.0
..   ...       ...   ...      ...      ...
693   38       138   175      173      0.0
694   47       138   257      156      0.0
695   67       115   564      160      1.6
696   38       138   175      173      0.0
699   57       128   303      159      0.0

[353 rows x 5 columns]

<div style="direction:rtl;line-height:300%;text-align:justify;" align="justify"><font face="B nazanin" size=4>
تابع 
normal_params
به عنوان ورودی یک دیتافریم می‌گیرد که تمام ستون‌های آن پیوسته هستند و در خروجی پارامترهای توزیع نرمال چند متغیری ستون‌ها را برمی‌گرداند
</font></div>

In [20]:
def normal_params(df):
    covMatrix = pd.DataFrame.cov(df)
    sigma=covMatrix.values
    sigma=np.matrix(sigma)
    mu=df.mean()
    mu=np.array(mu)

    return sigma,mu

In [21]:
sigma0,mu0=normal_params(data_base_0)

In [22]:
sigma0

matrix([[ 6.62505880e+01,  3.44397794e+01,  6.00793998e+01,
         -2.66701944e+01,  5.88020471e-01],
        [ 3.44397794e+01,  3.35708263e+02,  1.13904204e+02,
          1.14773179e+01,  2.64363261e+00],
        [ 6.00793998e+01,  1.13904204e+02,  2.49670714e+03,
          8.16417569e+01, -4.58391136e+00],
        [-2.66701944e+01,  1.14773179e+01,  8.16417569e+01,
          5.12423424e+02, -6.60629286e+00],
        [ 5.88020471e-01,  2.64363261e+00, -4.58391136e+00,
         -6.60629286e+00,  1.77429817e+00]])

In [23]:
mu0

array([ 56.51149425, 134.47413793, 248.86206897, 140.01436782,
         1.67672414])

In [24]:
sigma1,mu1=normal_params(data_base_1)

In [25]:
sigma1

matrix([[ 9.73707829e+01,  4.80951101e+01,  1.12282522e+02,
         -1.03489956e+02,  1.35330044e+00],
        [ 4.80951101e+01,  2.52732391e+02,  9.16842165e+01,
          2.81241952e-01,  1.86688369e+00],
        [ 1.12282522e+02,  9.16842165e+01,  2.65953963e+03,
         -6.05986834e+00,  1.67960099e+00],
        [-1.03489956e+02,  2.81241952e-01, -6.05986834e+00,
          3.50936566e+02, -2.98779858e+00],
        [ 1.35330044e+00,  1.86688369e+00,  1.67960099e+00,
         -2.98779858e+00,  6.18231554e-01]])

In [26]:
mu1

array([ 52.12464589, 128.83002833, 238.58073654, 159.28328612,
         0.59405099])

<div style="direction:rtl;line-height:300%;text-align:justify;" align="justify"><font face="B nazanin" size=4>
تابع زیر با گرفتن یک بردار از ویژگی‌های پیوسته و پارامتر‌های توزیع نرمال چند متغیره، احتمال دیده شدن آن بردار را برمی‌گرداند
</font></div>

<div style="direction:rtl;line-height:300%;text-align:justify;" align="justify"><font face="B nazanin" size=4>
تابع 
cat_params
با گرفتن یک دیتافریم و اسم یک ستون گسسته از آن احتمال دیده شدن هر یک از مقادیر منحصر به فرد آن ستون را برمی‌گرداند.
</font></div>

<div style="direction:rtl;line-height:300%;text-align:justify;" align="justify"><font face="B nazanin" size=4>
تابع زیر با گرفتن یک بردار از ویژگی‌های گسسته و یک دیتافریم، احتمال دیده شدن آن بردار را برمی‌گرداند.
</font></div>

<div style="direction:rtl;line-height:300%;text-align:justify;" align="justify"><font face="B nazanin" size=4>
تابع زیر با گرفتن یک دیتافریم، پارامتر‌های مورد نیاز برای توزیع های پیوسته و گسسته و یک بردار از ویژگی‌ها، احتمال دیده شدن آن بردار را در خروجی برمی‌گرداند.
</font></div>

<div style="direction:rtl;line-height:300%;text-align:justify;" align="justify"><font face="B nazanin" size=4>
<b style="font-size: 20px">ارزیابی (دستیار آموزشی)</b>
<br>
</font></div>

In [27]:
data_base_test = pd.read_csv("heart_test.csv")
print(data_base_test) 

     Unnamed: 0  age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  \
0             0   60    1   0       130   253    0        1      144      1   
1             1   67    1   0       160   286    0        0      108      1   
2             2   41    1   2       130   214    0        0      168      0   
3             3   68    1   2       180   274    1        0      150      1   
4             4   64    0   2       140   313    0        1      133      0   
..          ...  ...  ...  ..       ...   ...  ...      ...      ...    ...   
146         146   65    0   0       150   225    0        0      114      0   
147         147   46    0   1       105   204    0        1      172      0   
148         148   57    1   0       152   274    0        1       88      1   
149         149   55    1   1       130   262    0        1      155      0   
150         150   67    0   0       106   223    0        1      142      0   

     oldpeak  slope  ca  thal  target  
0        1.

In [28]:
data_base_test.drop(["Unnamed: 0"], axis = 1, inplace = True)
data_base_test

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     60    1   0       130   253    0        1      144      1      1.4   
1     67    1   0       160   286    0        0      108      1      1.5   
2     41    1   2       130   214    0        0      168      0      2.0   
3     68    1   2       180   274    1        0      150      1      1.6   
4     64    0   2       140   313    0        1      133      0      0.2   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
146   65    0   0       150   225    0        0      114      0      1.0   
147   46    0   1       105   204    0        1      172      0      0.0   
148   57    1   0       152   274    0        1       88      1      1.2   
149   55    1   1       130   262    0        1      155      0      0.0   
150   67    0   0       106   223    0        1      142      0      0.3   

     slope  ca  thal  target  
0        2   1     3       0  
1        1   3     2       0  
2        1   0     2       1  
3        1   0     3       0  
4        2   0     3       1  
..     ...  ..   ...     ...  
146      1   3     3       0  
147      2   0     2       1  
148      1   1     3       0  
149      2   0     2       1  
150      2   2     2       1  

[151 rows x 14 columns]

In [29]:
data_base_disceret_test = pd.DataFrame(data_base_test,columns=list_discrete_tag)
data_base_disceret_test

sex  cp  fbs  restecg  exang  slope  ca  thal  target
0      1   0    0        1      1      2   1     3       0
1      1   0    0        0      1      1   3     2       0
2      1   2    0        0      0      1   0     2       1
3      1   2    1        0      1      1   0     3       0
4      0   2    0        1      0      2   0     3       1
..   ...  ..  ...      ...    ...    ...  ..   ...     ...
146    0   0    0        0      0      1   3     3       0
147    0   1    0        1      0      2   0     2       1
148    1   0    0        1      1      1   1     3       0
149    1   1    0        1      0      2   0     2       1
150    0   0    0        1      0      2   2     2       1

[151 rows x 9 columns]

In [30]:
y_true=data_base_disceret_test['target']
y_true

0      0
1      0
2      1
3      0
4      1
      ..
146    0
147    1
148    0
149    1
150    1
Name: target, Length: 151, dtype: int64

In [31]:
data_base_disceret_test.drop('target',axis=1,inplace=True)
data_base_disceret_test

sex  cp  fbs  restecg  exang  slope  ca  thal
0      1   0    0        1      1      2   1     3
1      1   0    0        0      1      1   3     2
2      1   2    0        0      0      1   0     2
3      1   2    1        0      1      1   0     3
4      0   2    0        1      0      2   0     3
..   ...  ..  ...      ...    ...    ...  ..   ...
146    0   0    0        0      0      1   3     3
147    0   1    0        1      0      2   0     2
148    1   0    0        1      1      1   1     3
149    1   1    0        1      0      2   0     2
150    0   0    0        1      0      2   2     2

[151 rows x 8 columns]

In [32]:
for i in range(len(list_discrete_tag)):
    data_base_test.drop(list_discrete_tag[i],axis=1,inplace=True)
data_base_test

age  trestbps  chol  thalach  oldpeak
0     60       130   253      144      1.4
1     67       160   286      108      1.5
2     41       130   214      168      2.0
3     68       180   274      150      1.6
4     64       140   313      133      0.2
..   ...       ...   ...      ...      ...
146   65       150   225      114      1.0
147   46       105   204      172      0.0
148   57       152   274       88      1.2
149   55       130   262      155      0.0
150   67       106   223      142      0.3

[151 rows x 5 columns]

In [33]:
def Accuracy(y_true,y_pred):
    true=np.size(np.where(y_pred==y_true)) 
    acc=true/len(y_true)*100
    return acc

<div style="direction:rtl;line-height:300%;text-align:justify;" align="justify"><font face="B nazanin" size=4>
توابع زیر با گرفتن دو بردار که یکی همان 
target 
و دیگری همان  
prediction
است، معیارهای خواسته شده را محاسبه می‌کنند.
</font></div>

In [34]:
def cal_pred_detail(y_true,y_pred):
    tp=0
    tn=0
    fp=0
    fn=0
    for i in range (len(y_true)):
        if (y_pred[i]==y_true[i] and y_pred[i]==0):
            tn=tn+1
        elif (y_pred[i]==y_true[i] and y_pred[i]==1):
            tp=tp+1
        elif (y_pred[i]!=y_true[i] and y_pred[i]==0):
            fn=fn+1
        else:
            fp=fp+1
    return tp,tn,fp,fn
        
    

In [35]:
def precision(y_true, y_pred):
    tp,tn,fp,fn=cal_pred_detail(y_true,y_pred)
    p=tp/(tp+fp)*100
    return p

In [36]:
def recall(y_true, y_pred):
    tp,tn,fp,fn=cal_pred_detail(y_true,y_pred)
    r=tp/(tp+fn)*100
    return r

In [37]:
def f1score(y_true, y_pred):
    tp,tn,fp,fn=cal_pred_detail(y_true,y_pred)
    p=tp/(tp+fp)
    r=tp/(tp+fn)
    f=2*p*r/(p+r)*100
    
    return f

# 2

In [38]:
data_base = pd.read_csv("heart_train.csv")
print(data_base) 

     Unnamed: 0  age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  \
0             0   57    1   2       150   168    0        1      174      0   
1             1   46    1   1       101   197    1        1      156      0   
2             2   54    0   2       135   304    1        1      170      0   
3             3   48    1   1       110   229    0        1      168      0   
4             4   58    0   0       130   197    0        1      131      0   
..          ...  ...  ...  ..       ...   ...  ...      ...      ...    ...   
696         696   38    1   2       138   175    0        1      173      0   
697         697   47    1   2       108   243    0        1      152      0   
698         698   61    0   0       145   307    0        0      146      1   
699         699   57    0   0       128   303    0        0      159      0   
700         700   60    1   0       130   253    0        1      144      1   

     oldpeak  slope  ca  thal  target  
0        1.

In [39]:
data_base.head()

Unnamed: 0  age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  \
0           0   57    1   2       150   168    0        1      174      0   
1           1   46    1   1       101   197    1        1      156      0   
2           2   54    0   2       135   304    1        1      170      0   
3           3   48    1   1       110   229    0        1      168      0   
4           4   58    0   0       130   197    0        1      131      0   

   oldpeak  slope  ca  thal  target  
0      1.6      2   0     2       1  
1      0.0      2   0     3       1  
2      0.0      2   0     2       1  
3      1.0      0   0     3       0  
4      0.6      1   0     2       1

In [40]:
data_base.drop(["Unnamed: 0"], axis = 1, inplace = True)

In [41]:
data_base

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     57    1   2       150   168    0        1      174      0      1.6   
1     46    1   1       101   197    1        1      156      0      0.0   
2     54    0   2       135   304    1        1      170      0      0.0   
3     48    1   1       110   229    0        1      168      0      1.0   
4     58    0   0       130   197    0        1      131      0      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
696   38    1   2       138   175    0        1      173      0      0.0   
697   47    1   2       108   243    0        1      152      0      0.0   
698   61    0   0       145   307    0        0      146      1      1.0   
699   57    0   0       128   303    0        0      159      0      0.0   
700   60    1   0       130   253    0        1      144      1      1.4   

     slope  ca  thal  target  
0        2   0     2       1  
1        2   0     3       1  
2        2   0     2       1  
3        0   0     3       0  
4        1   0     2       1  
..     ...  ..   ...     ...  
696      2   4     2       1  
697      2   0     2       0  
698      1   0     3       0  
699      2   1     2       1  
700      2   1     3       0  

[701 rows x 14 columns]

In [42]:
header=data_base.columns
header

Index(['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach',
       'exang', 'oldpeak', 'slope', 'ca', 'thal', 'target'],
      dtype='object')

In [43]:
list_discrete_no=[]
list_discrete_tag=[]
for i in range(len(header)):
    if (len(np.unique(data_base[header[i]]))<=5):
        print(header[i])
        list_discrete_tag.append(header[i])
        print(np.unique(data_base[header[i]]))
        list_discrete_no.append(np.unique(data_base[header[i]]))
        

sex
[0 1]
cp
[0 1 2 3]
fbs
[0 1]
restecg
[0 1 2]
exang
[0 1]
slope
[0 1 2]
ca
[0 1 2 3 4]
thal
[0 1 2 3]
target
[0 1]


In [44]:
print(list_discrete_no)

[array([0, 1], dtype=int64), array([0, 1, 2, 3], dtype=int64), array([0, 1], dtype=int64), array([0, 1, 2], dtype=int64), array([0, 1], dtype=int64), array([0, 1, 2], dtype=int64), array([0, 1, 2, 3, 4], dtype=int64), array([0, 1, 2, 3], dtype=int64), array([0, 1], dtype=int64)]


In [45]:
list_discrete_tag

['sex', 'cp', 'fbs', 'restecg', 'exang', 'slope', 'ca', 'thal', 'target']

In [46]:
gp=data_base.groupby(by='target').size()
gp

target
0    348
1    353
dtype: int64

### 5

<div style="direction:ltr;line-height:300%;text-align:justify;" align="justify"><font face="B nazanin" size=5>
    <b>Logistic Regression</b>
</font></div>

In [47]:
from sklearn.model_selection import train_test_split

<div style="direction:rtl;line-height:300%;text-align:justify;" align="justify"><font face="B nazanin" size=4>
تابع زیر یک بردار وزن می‌گیرد و در خروجی رگولاریزیشن 
L2
را برمی‌گرداند
</font></div>

In [48]:
def l2(W):
    lr2=2*W
    return lr2

<div style="direction:rtl;line-height:300%;text-align:justify;" align="justify"><font face="B nazanin" size=4>
تابع زیر یک بردار می‌گیرد و سیگموید مولفه‌های آن را در خروجی برمی‌گرداند
</font></div>

In [49]:
def sigm(W,x,b):
    si = 1 / ( 1 + np.exp( - (f(W,x) + b ) ) ) 
    return si


<div style="direction:rtl;line-height:300%;text-align:justify;" align="justify"><font face="B nazanin" size=4>
تابع زیر با گرفتن یک بردار وزن و یک ماتریس از ویژگی‌ها، در خروجی نتیجه‌ی ضرب داخلی ویژگی‌ها و ماتریس را برمی‌گرداند.
</font></div>

In [50]:
def f(W, x):
    xw=x.dot( W )
    return xw

<div style="direction:rtl;line-height:300%;text-align:justify;" align="justify"><font face="B nazanin" size=4>
تابع زیر با گرفتن بردار وزن، ماتریس ویژگی‌ها، بردار هدف و لامبدا، گرادیان مورد نیاز را محاسبه می‌کند
</font></div>

In [51]:
def dj(W, X, y, lam,m):
    A=sigmoid()       
    tmp = ( A - Y.T )        
    tmp = np.reshape( tmp, m )        
    dW = (np.dot( X.T, tmp ) /m)+(lam*l2(W))
    db = np.sum( tmp ) / m 

    W = W - lr * dW     
    b = b - lr * db

    return self

<div style="direction:rtl;line-height:300%;text-align:justify;" align="justify"><font face="B nazanin" size=4>
تابع زیر خروجی تابع 
f
را می‌گیرد و مشخص می‌کند هر مولفه در کدام دسته قرار خواهد گرفت.
</font></div>

In [52]:
def classify(X,W,b):
    Z = 1 / ( 1 + np.exp( - ( X.dot( W ) + b ) ) )        
    Y = np.where( Z > 0.5, 1, 0 )        
    return Y
 

<div style="direction:rtl;line-height:300%;text-align:justify;" align="justify"><font face="B nazanin" size=4>
با استفاده از توابع گفته شده خواسته‌ی پنجم سوال را پاسخ دهید.
</font></div>

In [53]:
class LR() :
    def __init__( self, lr, iterations,lam ) :        
        self.lr = lr
        self.lam=lam
        self.iterations = iterations
              
    def f(self):
        xw=self.X.dot( self.W )
        return xw
    
    def l2(self):
        lr2=2*self.W
        return lr2
        
        
    def sigm(self):
        si = 1 / ( 1 + np.exp( - ( self.f() + self.b ) ) ) 
        return si
        
      
    def dj( self ) : 
        
        si=self.sigm()       
        tmp = ( si - self.Y.T )        
        tmp = np.reshape( tmp, self.m )        
        dW = (np.dot( self.X.T, tmp ) / self.m)+(self.lam*self.l2())
        db = np.sum( tmp ) / self.m 
            
        self.W = self.W - self.lr * dW     
        self.b = self.b - self.lr * db
          
        return self
    
    def fit( self, X, Y ) :        
        print(X.shape)
        print(Y.shape)
        self.m, self.n = X.shape                
        self.W = np.zeros( self.n )        
        self.b = 0        
        self.X = X        
        self.Y = Y
                  
        for i in range( self.iterations ) :            
            self.dj()            
        return self,self.W
      
 
      
    def classify( self, X ) :    
        Z = 1 / ( 1 + np.exp( - ( X.dot( self.W ) + self.b ) ) )        
        Y = np.where( Z > 0.5, 1, 0 )        
        return Y
  

In [54]:
df = pd.read_csv( "heart_train.csv")
df.drop(["Unnamed: 0"], axis = 1, inplace = True)
print(df)

     age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     57    1   2       150   168    0        1      174      0      1.6   
1     46    1   1       101   197    1        1      156      0      0.0   
2     54    0   2       135   304    1        1      170      0      0.0   
3     48    1   1       110   229    0        1      168      0      1.0   
4     58    0   0       130   197    0        1      131      0      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
696   38    1   2       138   175    0        1      173      0      0.0   
697   47    1   2       108   243    0        1      152      0      0.0   
698   61    0   0       145   307    0        0      146      1      1.0   
699   57    0   0       128   303    0        0      159      0      0.0   
700   60    1   0       130   253    0        1      144      1      1.4   

     slope  ca  thal  target  
0        2   0     2       1  
1        2   0     3     

In [55]:
X = df.iloc[:,:-1].values
Y = df.iloc[:,-1:].values
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, random_state = 0 )  

## lambda=10

In [56]:
model = LR( lr = 0.001, iterations = 100000,lam=10 )
model.fit( X_train, Y_train )      
Y_pred = model.classify( X_test ) 
Y_test=Y_test.flatten()
Accuracy(Y_test,Y_pred)

(560, 13)
(560, 1)


51.77304964539007

## lambda=1

In [59]:
model = LR( lr = 0.001, iterations = 100000,lam=1 )
model.fit( X_train, Y_train )      
Y_pred = model.classify( X_test ) 
Y_test=Y_test.flatten()
Accuracy(Y_test,Y_pred)

(560, 13)
(560, 1)


59.57446808510638

## lambda=0.1

In [58]:
model = LR( lr = 0.001, iterations = 100000,lam=0.1 )
model.fit( X_train, Y_train )      
Y_pred = model.classify( X_test ) 
Y_test=Y_test.flatten()
Accuracy(Y_test,Y_pred)

(560, 13)
(560, 1)


62.4113475177305

## lambda=0.01

In [60]:
model = LR( lr = 0.001, iterations = 100000,lam=0.01 )
model.fit( X_train, Y_train )      
Y_pred = model.classify( X_test ) 
Y_test=Y_test.flatten()
Accuracy(Y_test,Y_pred)

(560, 13)
(560, 1)


69.50354609929079

<div style="direction:rtl;line-height:300%;text-align:justify;" align="justify"><font face="B nazanin" size=4>
<b style="font-size: 20px">ارزیابی (دستیار آموزشی)</b>
</font></div>

In [61]:
test = pd.read_csv( "heart_test.csv")
test.drop(["Unnamed: 0"], axis = 1, inplace = True)
print(test)

     age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     60    1   0       130   253    0        1      144      1      1.4   
1     67    1   0       160   286    0        0      108      1      1.5   
2     41    1   2       130   214    0        0      168      0      2.0   
3     68    1   2       180   274    1        0      150      1      1.6   
4     64    0   2       140   313    0        1      133      0      0.2   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
146   65    0   0       150   225    0        0      114      0      1.0   
147   46    0   1       105   204    0        1      172      0      0.0   
148   57    1   0       152   274    0        1       88      1      1.2   
149   55    1   1       130   262    0        1      155      0      0.0   
150   67    0   0       106   223    0        1      142      0      0.3   

     slope  ca  thal  target  
0        2   1     3       0  
1        1   3     2     

In [62]:
X_test = test.iloc[:,:-1].values
Y_test = test.iloc[:,-1:].values
Y_test=Y_test.flatten()


In [63]:
def predict_test_TA(X_train, Y_train,X_test, Y_test,lamda):
    model = LR( lr = 0.001, iterations = 100000,lam=lamda )
    model.fit( X_train, Y_train )      
    Y_pred = model.classify(X_test) 
    return Accuracy(Y_test,Y_pred)
    

In [64]:
predict_test_TA(X_train, Y_train,X_test, Y_test,0.01)

(560, 13)
(560, 1)


66.22516556291392

<div style="direction:ltr;line-height:300%;text-align:justify;" align="justify"><font face="B nazanin" size=5>
    <b style="font-size: 35px">svm and kernel</b>
</font></div>

In [65]:
from sklearn.svm import SVC
from sklearn.model_selection import train_test_split

### 1

In [66]:
train = pd.read_csv("heart_train.csv")
print(train) 

     Unnamed: 0  age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  \
0             0   57    1   2       150   168    0        1      174      0   
1             1   46    1   1       101   197    1        1      156      0   
2             2   54    0   2       135   304    1        1      170      0   
3             3   48    1   1       110   229    0        1      168      0   
4             4   58    0   0       130   197    0        1      131      0   
..          ...  ...  ...  ..       ...   ...  ...      ...      ...    ...   
696         696   38    1   2       138   175    0        1      173      0   
697         697   47    1   2       108   243    0        1      152      0   
698         698   61    0   0       145   307    0        0      146      1   
699         699   57    0   0       128   303    0        0      159      0   
700         700   60    1   0       130   253    0        1      144      1   

     oldpeak  slope  ca  thal  target  
0        1.

In [67]:
train.drop(["Unnamed: 0"], axis = 1, inplace = True)
train

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     57    1   2       150   168    0        1      174      0      1.6   
1     46    1   1       101   197    1        1      156      0      0.0   
2     54    0   2       135   304    1        1      170      0      0.0   
3     48    1   1       110   229    0        1      168      0      1.0   
4     58    0   0       130   197    0        1      131      0      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
696   38    1   2       138   175    0        1      173      0      0.0   
697   47    1   2       108   243    0        1      152      0      0.0   
698   61    0   0       145   307    0        0      146      1      1.0   
699   57    0   0       128   303    0        0      159      0      0.0   
700   60    1   0       130   253    0        1      144      1      1.4   

     slope  ca  thal  target  
0        2   0     2       1  
1        2   0     3       1  
2        2   0     2       1  
3        0   0     3       0  
4        1   0     2       1  
..     ...  ..   ...     ...  
696      2   4     2       1  
697      2   0     2       0  
698      1   0     3       0  
699      2   1     2       1  
700      2   1     3       0  

[701 rows x 14 columns]

In [68]:
test = pd.read_csv("heart_test.csv")
print(test) 

     Unnamed: 0  age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  \
0             0   60    1   0       130   253    0        1      144      1   
1             1   67    1   0       160   286    0        0      108      1   
2             2   41    1   2       130   214    0        0      168      0   
3             3   68    1   2       180   274    1        0      150      1   
4             4   64    0   2       140   313    0        1      133      0   
..          ...  ...  ...  ..       ...   ...  ...      ...      ...    ...   
146         146   65    0   0       150   225    0        0      114      0   
147         147   46    0   1       105   204    0        1      172      0   
148         148   57    1   0       152   274    0        1       88      1   
149         149   55    1   1       130   262    0        1      155      0   
150         150   67    0   0       106   223    0        1      142      0   

     oldpeak  slope  ca  thal  target  
0        1.

In [69]:
test.drop(["Unnamed: 0"], axis = 1, inplace = True)
test

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     60    1   0       130   253    0        1      144      1      1.4   
1     67    1   0       160   286    0        0      108      1      1.5   
2     41    1   2       130   214    0        0      168      0      2.0   
3     68    1   2       180   274    1        0      150      1      1.6   
4     64    0   2       140   313    0        1      133      0      0.2   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
146   65    0   0       150   225    0        0      114      0      1.0   
147   46    0   1       105   204    0        1      172      0      0.0   
148   57    1   0       152   274    0        1       88      1      1.2   
149   55    1   1       130   262    0        1      155      0      0.0   
150   67    0   0       106   223    0        1      142      0      0.3   

     slope  ca  thal  target  
0        2   1     3       0  
1        1   3     2       0  
2        1   0     2       1  
3        1   0     3       0  
4        2   0     3       1  
..     ...  ..   ...     ...  
146      1   3     3       0  
147      2   0     2       1  
148      1   1     3       0  
149      2   0     2       1  
150      2   2     2       1  

[151 rows x 14 columns]

In [70]:
X=train.drop('target',axis=1)
Y=train['target']

In [71]:
x_train,x_val,y_train,y_val=train_test_split(X,Y,test_size=0.2,random_state=42)

In [72]:
x_test=test.drop('target',axis=1)
y_test=test['target']

## C=0.1

In [73]:
SVMClassi=SVC(C=0.1,kernel='linear')
SVMClassi.fit(x_train,y_train)
y_pred=SVMClassi.predict(x_val)
Accuracy(y_val,y_pred)

86.52482269503547

## C=0.01

In [74]:
SVMClassi=SVC(C=0.01,kernel='linear')
SVMClassi.fit(x_train,y_train)
y_pred=SVMClassi.predict(x_val)
Accuracy(y_val,y_pred)

83.68794326241135

## C=0.001

In [75]:
SVMClassi=SVC(C=0.001,kernel='linear')
SVMClassi.fit(x_train,y_train)
y_pred=SVMClassi.predict(x_val)
Accuracy(y_val,y_pred)

71.63120567375887

## C=0.0001

In [76]:
SVMClassi=SVC(C=0.0001,kernel='linear')
SVMClassi.fit(x_train,y_train)
y_pred=SVMClassi.predict(x_val)
Accuracy(y_val,y_pred)

70.92198581560284

<div style="direction:rtl;line-height:300%;text-align:justify;" align="justify"><font face="B nazanin" size=4>
<b style="font-size: 20px">ارزیابی (دستیار آموزشی)</b>
</font></div>

In [77]:
def predict_test_TA(x_train,y_train,x_test,y_test,c):
    SVMClassi=SVC(C=c,kernel='linear')
    SVMClassi.fit(x_train,y_train)
    y_pred=SVMClassi.predict(x_test)
    print(Accuracy(y_test,y_pred))
    

In [78]:
predict_test_TA(X,Y,x_test,y_test,0.1)

80.79470198675497


<div style="direction:rtl;line-height:300%;text-align:justify;" align="justify"><font face="B nazanin" size=4>
<b style="font-size: 20px">با توجه به اعتبارسنجی که انجام شد مقدار 0.1 به عنوان بهترین مقدار انتخاب شد و این مسئله می تواند به این سبب باشد که با انتخاب مقدار بزرگ برای این پارامتر به دنبال کاهش حطای آموزش هستیم و تا حدی از افزایش بردارهای حاشیه صرف نظر می کنیم بنابراین این مسئله سبب افزایش دقت می شود زیرا همانطور گفته شد از افزایش مقدار حاشیه ها جلوگیری کرده ایم. </b>
</font></div>

### 2

## gamma=0.001

In [79]:
SVMClassi=SVC(C=1,kernel='rbf',gamma=0.001)
SVMClassi.fit(x_train,y_train)
y_pred=SVMClassi.predict(x_val)
f1score(np.array(y_val),np.array(y_pred))

79.13669064748201

## gamma=0.0001

In [80]:
SVMClassi=SVC(C=1,kernel='rbf',gamma=0.0001)
SVMClassi.fit(x_train,y_train)
y_pred=SVMClassi.predict(x_val)
f1score(np.array(y_val),np.array(y_pred))

66.1764705882353

## gamma=0.00001

In [81]:
SVMClassi=SVC(C=1,kernel='rbf',gamma=0.00001)
SVMClassi.fit(x_train,y_train)
y_pred=SVMClassi.predict(x_val)
f1score(np.array(y_val),np.array(y_pred))

70.4225352112676

<div style="direction:rtl;line-height:300%;text-align:justify;" align="justify"><font face="B nazanin" size=4>
<b style="font-size: 20px">با افزایش مقدار گاما مرز دسته بندی ما پیچیده تر می شود و این مسئله می تواند سبب بیش برازش شود.</b>
</font></div>

<div style="direction:rtl;line-height:300%;text-align:justify;" align="justify"><font face="B nazanin" size=4>
<b style="font-size: 20px">ارزیابی (دستیار آموزشی)</b>
</font></div>

In [82]:
def predict_test_TA(x_train,y_train,x_test,y_test,g):
    SVMClassi=SVC(C=1,kernel='rbf',gamma=g)
    SVMClassi.fit(x_train,y_train)
    y_pred=SVMClassi.predict(x_test)
    return f1score(np.array(y_test),np.array(y_pred))
    

In [83]:
predict_test_TA(X,Y,x_test,y_test,0.001)

79.01234567901234

<div style="direction:ltr;line-height:300%;text-align:justify;" align="justify"><font face="B nazanin" size=4>
<b style="font-size: 35px">KNN</b>
</font></div>

In [84]:
import numpy as np
import pandas as pd
from scipy.stats import mode
import matplotlib.pyplot as plt

%matplotlib inline

<div style="direction:rtl;line-height:300%;text-align:justify;" align="justify"><font face="B nazanin" size=4>
تابع زیر دو بردار می‌گیرد و در خروجی فاصله‌ی اقلیدسی آن‌ها را برمی‌گرداند
</font></div>

In [85]:
def euclidean(a,b):
    return np.linalg.norm(a-b)

<div style="direction:rtl;line-height:300%;text-align:justify;" align="justify"><font face="B nazanin" size=4>
تابع زیر یک بردار و یک دیتافریم می‌گیرد و فاصله‌ی آن بردار با تمام بردار‌های موجود در دیتافریم را برمی‌گرداند
</font></div>

In [86]:
def distance(x,m,X_train,Y_train,k):
    euclidean_distances = np.zeros(m)
    for i in range(m) :
        d = euclidean( x, X_train[i] )
        euclidean_distances[i] = d
    inds = euclidean_distances.argsort()
    Y_train_sorted = Y_train[inds]

    return Y_train_sorted[:K]


<div style="direction:rtl;line-height:300%;text-align:justify;" align="justify"><font face="B nazanin" size=4>
تابع زیر یک بردار، یک دیتافریم و یک عدد طبیعی 
n
می‌گیرد و 
n
نزدیک ترین همسایه از آن بردار را در خروجی برمی‌گرداند
</font></div>

In [87]:
def nearest(X_test,k):
    m_test, n = X_test.shape
    Y_nearest = np.zeros( m_test )

    for i in range( m_test) :
        x = X_test[i]
        neighbors = np.zeros(K)
        neighbors = distance(x)
        Y_nearest[i] = knn_decide(neighbors)  
    return Y_nearest

<div style="direction:rtl;line-height:300%;text-align:justify;" align="justify"><font face="B nazanin" size=4>
تابع زیر با گرفتن همان ورودی‌های قبلی، این بار بردار 
v
را دسته بندی می‌کند.
</font></div>

In [88]:
def knn_decide(neighbors):
    return mode(neighbors)[0][0]  

In [89]:
class KNN() : 
      
    def __init__( self, K ) :
        self.K = K
          
    def fit( self, X_train, Y_train ) :
        self.X_train = X_train
        self.Y_train = Y_train
        self.m, self.n = X_train.shape
        
    def distance( self, x ) :
        euclidean_distances = np.zeros( self.m )
        for i in range( self.m ) :
            d = self.euclidean( x, self.X_train[i] )
            euclidean_distances[i] = d
        inds = euclidean_distances.argsort()
        Y_train_sorted = self.Y_train[inds]
        return Y_train_sorted[:self.K]
    
    def euclidean( self, x, x_train ):
          
        return np.linalg.norm(x-x_train)
    
    def knn_decide(self, neighbors):
         return mode( neighbors )[0][0]  
        
    def nearest( self, X_test ) :
        self.X_test = X_test
        self.m_test, self.n = X_test.shape
        Y_nearest = np.zeros( self.m_test )
        for i in range( self.m_test ) :
            x = self.X_test[i]
            neighbors = np.zeros( self.K )
            neighbors = self.distance( x )
            Y_nearest[i] = self.knn_decide(neighbors)  
              
        return Y_nearest

In [90]:
train = pd.read_csv("heart_train.csv")
print(train) 

     Unnamed: 0  age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  \
0             0   57    1   2       150   168    0        1      174      0   
1             1   46    1   1       101   197    1        1      156      0   
2             2   54    0   2       135   304    1        1      170      0   
3             3   48    1   1       110   229    0        1      168      0   
4             4   58    0   0       130   197    0        1      131      0   
..          ...  ...  ...  ..       ...   ...  ...      ...      ...    ...   
696         696   38    1   2       138   175    0        1      173      0   
697         697   47    1   2       108   243    0        1      152      0   
698         698   61    0   0       145   307    0        0      146      1   
699         699   57    0   0       128   303    0        0      159      0   
700         700   60    1   0       130   253    0        1      144      1   

     oldpeak  slope  ca  thal  target  
0        1.

In [91]:
train.drop(["Unnamed: 0"], axis = 1, inplace = True)
train

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     57    1   2       150   168    0        1      174      0      1.6   
1     46    1   1       101   197    1        1      156      0      0.0   
2     54    0   2       135   304    1        1      170      0      0.0   
3     48    1   1       110   229    0        1      168      0      1.0   
4     58    0   0       130   197    0        1      131      0      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
696   38    1   2       138   175    0        1      173      0      0.0   
697   47    1   2       108   243    0        1      152      0      0.0   
698   61    0   0       145   307    0        0      146      1      1.0   
699   57    0   0       128   303    0        0      159      0      0.0   
700   60    1   0       130   253    0        1      144      1      1.4   

     slope  ca  thal  target  
0        2   0     2       1  
1        2   0     3       1  
2        2   0     2       1  
3        0   0     3       0  
4        1   0     2       1  
..     ...  ..   ...     ...  
696      2   4     2       1  
697      2   0     2       0  
698      1   0     3       0  
699      2   1     2       1  
700      2   1     3       0  

[701 rows x 14 columns]

In [92]:
test = pd.read_csv("heart_test.csv")
print(test) 

     Unnamed: 0  age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  \
0             0   60    1   0       130   253    0        1      144      1   
1             1   67    1   0       160   286    0        0      108      1   
2             2   41    1   2       130   214    0        0      168      0   
3             3   68    1   2       180   274    1        0      150      1   
4             4   64    0   2       140   313    0        1      133      0   
..          ...  ...  ...  ..       ...   ...  ...      ...      ...    ...   
146         146   65    0   0       150   225    0        0      114      0   
147         147   46    0   1       105   204    0        1      172      0   
148         148   57    1   0       152   274    0        1       88      1   
149         149   55    1   1       130   262    0        1      155      0   
150         150   67    0   0       106   223    0        1      142      0   

     oldpeak  slope  ca  thal  target  
0        1.

In [93]:
test.drop(["Unnamed: 0"], axis = 1, inplace = True)
test

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     60    1   0       130   253    0        1      144      1      1.4   
1     67    1   0       160   286    0        0      108      1      1.5   
2     41    1   2       130   214    0        0      168      0      2.0   
3     68    1   2       180   274    1        0      150      1      1.6   
4     64    0   2       140   313    0        1      133      0      0.2   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
146   65    0   0       150   225    0        0      114      0      1.0   
147   46    0   1       105   204    0        1      172      0      0.0   
148   57    1   0       152   274    0        1       88      1      1.2   
149   55    1   1       130   262    0        1      155      0      0.0   
150   67    0   0       106   223    0        1      142      0      0.3   

     slope  ca  thal  target  
0        2   1     3       0  
1        1   3     2       0  
2        1   0     2       1  
3        1   0     3       0  
4        2   0     3       1  
..     ...  ..   ...     ...  
146      1   3     3       0  
147      2   0     2       1  
148      1   1     3       0  
149      2   0     2       1  
150      2   2     2       1  

[151 rows x 14 columns]

In [94]:
X = train.iloc[:,:-1].values
Y = train.iloc[:,-1:].values
X_test = test.iloc[:,:-1].values
Y_test = test.iloc[:,-1:].values


In [95]:
X_train, X_val, Y_train, Y_val = train_test_split(X, Y, test_size = 0.2, random_state = 0 )

### k = 1

In [96]:
KnnClassi = KNN(K = 1)
KnnClassi.fit( X_train, Y_train )
Y_pred = KnnClassi.nearest(X_val)
Y_val=Y_val.flatten()
Accuracy(Y_val,Y_pred)

91.48936170212765

### k = 5

In [97]:
KnnClassi = KNN(K =5)
KnnClassi.fit( X_train, Y_train )
Y_pred = KnnClassi.nearest( X_val )
Y_val=Y_val.flatten()
Accuracy(Y_val,Y_pred)

70.2127659574468

### k = 25

In [98]:
KnnClassi = KNN(K = 25)
KnnClassi.fit( X_train, Y_train )
Y_pred = KnnClassi.nearest( X_val )
Y_val=Y_val.flatten()
Accuracy(Y_val,Y_pred)

65.24822695035462

### k = 75

In [99]:
KnnClassi = KNN(K =75)
KnnClassi.fit( X_train, Y_train )
Y_pred = KnnClassi.nearest( X_val )
Y_val=Y_val.flatten()
Accuracy(Y_val,Y_pred)

62.4113475177305

<div style="direction:rtl;line-height:300%;text-align:justify;" align="justify"><font face="B nazanin" size=4>
<b style="font-size: 20px">ارزیابی (دستیار آموزشی)</b>
</font></div>

In [100]:
def predict_test_TA(X_train,Y_train,X_test,Y_test,k):
    KnnClassi = KNN(K=k)
    KnnClassi.fit( X_train, Y_train )
    Y_pred = KnnClassi.nearest( X_test)
    Y_test=Y_test.flatten()
    return Accuracy(Y_test,Y_pred)

In [101]:
predict_test_TA(X_train,Y_train,X_test,Y_test,1)

93.37748344370861

<div style="direction:ltr;line-height:300%;text-align:justify;" align="justify"><font face="B nazanin" size=4>
<b style="font-size: 35px">Ensemble Learning</b>
<br>
</font></div>

In [102]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import copy
from sklearn.utils import resample
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn import base
%matplotlib inline

<div style="direction:rtl;line-height:300%;text-align:justify;" align="justify"><font face="B nazanin" size=2>
<b style="font-size: 20px">زمانی که مقدار خطای وزنها یعنی (E) به مقدار نیم می رسد از این مرحله به بعد اضافه کردن درخت ها دیگر منجر به بهبود نمی شود  E هایی که بیشتر از 0.5 هستند به صورت تصادفی عمل می کند   </b>
</font></div>

### 1

<div style="direction:ltr;line-height:300%;text-align:justify;" align="justify"><font face="B nazanin" size=4>
<b style="font-size: 35px">Neural Network</b>
<br>
</font></div>

In [103]:
#Dependencies
import numpy as np
import pandas as pd
import sklearn
from sklearn.utils import shuffle
from sklearn.model_selection import train_test_split
import keras
from keras.models import Sequential
from keras.layers import Dense

<div style="direction:rtl;line-height:300%;text-align:justify;" align="justify"><font face="B nazanin" size=4>
تابع زیر دیتافریم می‌گیرد , پس از انجام پیش‌پردازش‌های لازم و تقسیم‌بندی به دو دسته,داده‌های آموزش و اعتبارسنجی را برمی‌گرداند
</font></div>

In [104]:
train = pd.read_csv("heart_train.csv")
train.drop(["Unnamed: 0"], axis = 1, inplace = True)
train

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     57    1   2       150   168    0        1      174      0      1.6   
1     46    1   1       101   197    1        1      156      0      0.0   
2     54    0   2       135   304    1        1      170      0      0.0   
3     48    1   1       110   229    0        1      168      0      1.0   
4     58    0   0       130   197    0        1      131      0      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
696   38    1   2       138   175    0        1      173      0      0.0   
697   47    1   2       108   243    0        1      152      0      0.0   
698   61    0   0       145   307    0        0      146      1      1.0   
699   57    0   0       128   303    0        0      159      0      0.0   
700   60    1   0       130   253    0        1      144      1      1.4   

     slope  ca  thal  target  
0        2   0     2       1  
1        2   0     3       1  
2        2   0     2       1  
3        0   0     3       0  
4        1   0     2       1  
..     ...  ..   ...     ...  
696      2   4     2       1  
697      2   0     2       0  
698      1   0     3       0  
699      2   1     2       1  
700      2   1     3       0  

[701 rows x 14 columns]

In [105]:
test = pd.read_csv("heart_test.csv")
test.drop(["Unnamed: 0"], axis = 1, inplace = True)
test

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     60    1   0       130   253    0        1      144      1      1.4   
1     67    1   0       160   286    0        0      108      1      1.5   
2     41    1   2       130   214    0        0      168      0      2.0   
3     68    1   2       180   274    1        0      150      1      1.6   
4     64    0   2       140   313    0        1      133      0      0.2   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
146   65    0   0       150   225    0        0      114      0      1.0   
147   46    0   1       105   204    0        1      172      0      0.0   
148   57    1   0       152   274    0        1       88      1      1.2   
149   55    1   1       130   262    0        1      155      0      0.0   
150   67    0   0       106   223    0        1      142      0      0.3   

     slope  ca  thal  target  
0        2   1     3       0  
1        1   3     2       0  
2        1   0     2       1  
3        1   0     3       0  
4        2   0     3       1  
..     ...  ..   ...     ...  
146      1   3     3       0  
147      2   0     2       1  
148      1   1     3       0  
149      2   0     2       1  
150      2   2     2       1  

[151 rows x 14 columns]

In [106]:
x_test = test.iloc[:,:-1].values
y_test = test.iloc[:,-1:].values

In [107]:
y_test=y_test.flatten()
X = train.iloc[:,:-1].values
Y = train.iloc[:,-1:].values

In [108]:
def preprocessing(train):
    X = train.iloc[:,:-1].values
    Y = train.iloc[:,-1:].values
    X_train, X_val, Y_train, Y_val = train_test_split(X, Y, test_size = 0.2, random_state = 0 )
    return X_train, Y_train, X_val, Y_val

<div style="direction:rtl;line-height:300%;text-align:justify;" align="justify"><font face="B nazanin" size=4>
تابع زیر داده‌ی آموزش و هایپرپارامتر‌های موردنیاز را (که به صورت دیکشنری است و یکی از کلید‌های آن layers بوده که مقدار آن  آرایه‌ای است شامل پارامتر‌های مربوط به هر لایه)را می‌گیرد
    و پس از پیاده‌سازی مدل و آموزش آن, مدل ترین‌شده را بازمی‌گرداند
</font></div>

In [109]:
def train_model(X_train, y_train, hparams):
    model = Sequential()
    model.add(Dense(hparams[0], input_dim=13, kernel_initializer='random_uniform', activation='relu'))
    model.add(Dense(hparams[1], input_dim=13, kernel_initializer='random_uniform', activation='relu'))
    model.add(Dense(hparams[2], input_dim=13, kernel_initializer='random_uniform', activation='sigmoid'))
    model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
    model.fit(X_train,y_train,epochs=100,batch_size=10)
    return model

<div style="direction:rtl;line-height:300%;text-align:justify;" align="justify"><font face="B nazanin" size=4>
تابع زیر در ورودی, مدل ترین‌شده و داده‌ی تست(یا اعتبارسنجی) را می‌گیرد و در خروجی, دقت مدل را برمی‌گرداند. از خروجی این تابع می‌توانید برای بررسی عملکرد مدل  استفاده کنید.</font></div>

In [110]:
def test_model(model, X_test, y_test):
    scores = model.evaluate(X_test,y_test)
    return scores[1]

<div style="direction:rtl;line-height:300%;text-align:justify;" align="justify"><font face="B nazanin" size=4>
حال با استفاده از توابع بالا کل فرآیند را در ذیل توسعه دهید.
</font></div>

In [111]:
hparams=[[15,7,1],[50,25,1],[100,50,1]]
best_hparam = []
val_accs = [] #compare results based on validation set
X_train, y_train, X_val, y_val=preprocessing(train)
for param in hparams:
    model=train_model(X_train,y_train,param)
    print('***************Validation*******************')
    scores = model.evaluate(X_val,y_val)
    print('********************************************')
    val_accs.append(scores)
    #test_acc = test_model(model, x_test, y_test)
    #print("Model Test Accuracy: {}".format(test_acc))

for i in range (len(val_accs)):
    if (val_accs[i][1]==np.max(val_accs)):
        best_hparam=hparams[i]
        

Epoch 1/100
56/56 [==============================] - 2s 4ms/step - loss: 0.6930 - accuracy: 0.4781
Epoch 2/100
56/56 [==============================] - 0s 4ms/step - loss: 0.6842 - accuracy: 0.6240
Epoch 3/100
56/56 [==============================] - 0s 4ms/step - loss: 0.6680 - accuracy: 0.6220
Epoch 4/100
56/56 [==============================] - 0s 3ms/step - loss: 0.6096 - accuracy: 0.7011
Epoch 5/100
56/56 [==============================] - 0s 3ms/step - loss: 0.5563 - accuracy: 0.7063
Epoch 6/100
56/56 [==============================] - 0s 3ms/step - loss: 0.5840 - accuracy: 0.6846
Epoch 7/100
56/56 [==============================] - 0s 3ms/step - loss: 0.5163 - accuracy: 0.7359
Epoch 8/100
56/56 [==============================] - 0s 3ms/step - loss: 0.5226 - accuracy: 0.7146
Epoch 9/100
56/56 [==============================] - 0s 3ms/step - loss: 0.4710 - accuracy: 0.7821
Epoch 10/100
56/56 [==============================] - 0s 3ms/step - loss: 0.4797 - accuracy: 0.7545
Epoch 11/

56/56 [==============================] - 0s 3ms/step - loss: 0.3268 - accuracy: 0.8497
Epoch 84/100
56/56 [==============================] - 0s 2ms/step - loss: 0.3633 - accuracy: 0.8279: 0s - loss: 0.3658 - accuracy: 0.
Epoch 85/100
56/56 [==============================] - 0s 3ms/step - loss: 0.3340 - accuracy: 0.8534
Epoch 86/100
56/56 [==============================] - 0s 3ms/step - loss: 0.3890 - accuracy: 0.8292
Epoch 87/100
56/56 [==============================] - 0s 3ms/step - loss: 0.3491 - accuracy: 0.8545
Epoch 88/100
56/56 [==============================] - 0s 3ms/step - loss: 0.3203 - accuracy: 0.8636
Epoch 89/100
56/56 [==============================] - 0s 3ms/step - loss: 0.3411 - accuracy: 0.8414
Epoch 90/100
56/56 [==============================] - 0s 2ms/step - loss: 0.3562 - accuracy: 0.8584
Epoch 91/100
56/56 [==============================] - 0s 2ms/step - loss: 0.3213 - accuracy: 0.8747
Epoch 92/100
56/56 [==============================] - 0s 4ms/step - loss: 0.357

56/56 [==============================] - 0s 3ms/step - loss: 0.3832 - accuracy: 0.8179
Epoch 42/100
56/56 [==============================] - 0s 3ms/step - loss: 0.3094 - accuracy: 0.8714
Epoch 43/100
56/56 [==============================] - 0s 3ms/step - loss: 0.3416 - accuracy: 0.8392
Epoch 44/100
56/56 [==============================] - 0s 3ms/step - loss: 0.3665 - accuracy: 0.8166
Epoch 45/100
56/56 [==============================] - 0s 3ms/step - loss: 0.4069 - accuracy: 0.8065
Epoch 46/100
56/56 [==============================] - 0s 2ms/step - loss: 0.3999 - accuracy: 0.8150
Epoch 47/100
56/56 [==============================] - 0s 2ms/step - loss: 0.3774 - accuracy: 0.8288
Epoch 48/100
56/56 [==============================] - 0s 3ms/step - loss: 0.3790 - accuracy: 0.8318
Epoch 49/100
56/56 [==============================] - 0s 3ms/step - loss: 0.3531 - accuracy: 0.8346
Epoch 50/100
56/56 [==============================] - 0s 2ms/step - loss: 0.3643 - accuracy: 0.8220
Epoch 51/100


<div style="direction:rtl;line-height:300%;text-align:justify;" align="justify"><font face="B nazanin" size=4>
<b style="font-size: 20px">ارزیابی (دستیار آموزشی)</b>
<br>
    کد شما در این بخش ارزیابی می‌شود. در این قسمت فارغ از دانستن داده‌ها, پس از طی کردن مراحل ترین و تست, مدل با بهترین نتایج را به تابع زیر پاس دهید و دقت آنرا چاپ کنید. 
</font></div>

In [112]:
def predict_test_TA(X,Y,best_hparam,x_test,y_test):
    model=train_model(X,Y,best_hparam)
    test_acc = test_model(model, x_test, y_test)
    return test_acc

In [113]:
predict_test_TA(X,Y,best_hparam,x_test,y_test)

Epoch 1/100
71/71 [==============================] - 3s 4ms/step - loss: 0.6929 - accuracy: 0.5105
Epoch 2/100
71/71 [==============================] - 0s 4ms/step - loss: 0.6495 - accuracy: 0.6422
Epoch 3/100
71/71 [==============================] - 0s 4ms/step - loss: 0.5842 - accuracy: 0.7053
Epoch 4/100
71/71 [==============================] - 0s 4ms/step - loss: 0.5586 - accuracy: 0.7114
Epoch 5/100
71/71 [==============================] - 0s 4ms/step - loss: 0.5600 - accuracy: 0.7098
Epoch 6/100
71/71 [==============================] - 0s 5ms/step - loss: 0.4810 - accuracy: 0.7450
Epoch 7/100
71/71 [==============================] - 0s 4ms/step - loss: 0.4820 - accuracy: 0.7671
Epoch 8/100
71/71 [==============================] - 0s 4ms/step - loss: 0.5093 - accuracy: 0.7528
Epoch 9/100
71/71 [==============================] - 0s 3ms/step - loss: 0.4167 - accuracy: 0.8036
Epoch 10/100
71/71 [==============================] - 0s 3ms/step - loss: 0.3987 - accuracy: 0.8145
Epoch 11/

0.8278145790100098

<div style="direction:rtl;line-height:300%;text-align:justify;" align="justify"><font face="B nazanin" size=4>
<b style="font-size: 30px">کلام آخر</b>
<br>
 در همین قسمت, نتایج الگوریتم‌های مختلفی را که توسعه دادید تنها براساس معیار Accuracy مقایسه کنید.
 و در مورد نتایج در حد یک پاراگراف حداکثر ۶ خطی (با همین فونت) بحث کنید. از دید شما برای این داده بهترین روش چیست؟ چرا؟
</font></div>

<div style="direction:rtl;line-height:300%;text-align:justify;" align="justify"><font face="B nazanin" size=4>
<b style="font-size: 15px">دقت بر روی دسته بند Logistic Regression</b>
<br>
66.22516556291392
<br>
<b style="font-size: 15px">دقت بر روی دسته بند Linear SVM</b>
<br>
80.79470198675497
<br>
<b style="font-size: 15px">دقت بر روی دسته بند rbf SVM</b>
<br>
79.01234567901234
<br>
<b style="font-size: 15px">دقت بر روی دسته بند KNN</b>
<br>
93.37748344370861
<br>
<b style="font-size: 15px">دقت بر روی دسته بند Neural Network</b>
<br>
0.8013244867324829
<br>
 روش های ensemble Learning بر روی این داده به نتایج بهتری می رسند. این مسئله تا حدی قابل انتظار است زیرا پایه ی این روش ها بر اساس پیش بنی جمعی است و چون بر اساس تعدادی زیادی دسته بندی پیش بنی انجام می شود دقت در دو روش random forest و adaboost نسبت به دو روش دیگر بالاتر است. اما همانطور که می دانیم اینکه الگوریتم های ensemble Learning به صورت کلی ارجحیتی نسبت به سایر الگوریتم ها ندارن . چه بسا اگر هایپر پارامتر های سایر الگوریتم ها مثلا شبکه ی عصبی به خوبی تنظیم شود، می تواند نتیجه ی بهتری حتی نسبت به الگوریتم های ensemble learning بگیرند.همچنین بعضی از اللگوریتم ها نیز به پیش پردازش داده ها حساس هستند و  ممکن است اگر داده های آموزش را پیش پردازش می کردیم بر روی  برخی از الگوریتم ها که نتیجه ی کم گرفته ایم نتیجه بهتری بگیریم. 
</font></div>